File: frontend/src/api/edgeAgents.ts
import apiClient from "./client";

export interface EdgeAgent {
  id: string;
  name: string;
  status: "online" | "offline" | "stale";
  version: string | null;
  last_seen: string | null;
  ip_address: string | null;
  device_ids: string[] | null;
  notes: string | null;
  created_at: string;
}

export interface EdgeAgentWithToken extends EdgeAgent {
  token: string;
}

export interface SsoConfig {
  id: string;
  provider: string;
  client_id: string;
  discovery_url: string;
  group_claim: string | null;
  group_mapping: Record<string, string> | null;
  sso_required: boolean;
  is_active: boolean;
  created_at: string;
}

export interface MarketplacePlugin {
  id: string;
  name: string;
  slug: string;
  version: string;
  category: string;
  description: string | null;
  is_builtin: boolean;
  download_count: number;
  approved_at: string | null;
}

export interface TenantPlugin {
  id: string;
  plugin_id: string;
  installed_at: string;
  plugin: MarketplacePlugin;
}

export interface RbacCustomRole {
  id: string;
  name: string;
  description: string | null;
  permissions: Record<string, unknown> | null;
  created_at: string;
}

export interface RbacRoleAssignment {
  id: string;
  user_id: string;
  role_id: string;
  assigned_at: string;
}

const BASE = "/platform";

export const edgeAgentsApi = {
  // Edge Agents
  listAgents: () =>
    apiClient.get<EdgeAgent[]>(`${BASE}/agents`).then(r => r.data),
  createAgent: (data: { name: string; notes?: string }) =>
    apiClient.post<EdgeAgentWithToken>(`${BASE}/agents`, data).then(r => r.data),
  updateAgent: (id: string, data: { name: string; notes?: string }) =>
    apiClient.patch<EdgeAgent>(`${BASE}/agents/${id}`, data).then(r => r.data),
  deleteAgent: (id: string) => apiClient.delete(`${BASE}/agents/${id}`),

  // SSO
  getSsoConfig: () =>
    apiClient.get<SsoConfig | null>(`${BASE}/sso`).then(r => r.data),
  upsertSsoConfig: (data: { provider?: string; client_id: string; client_secret?: string; discovery_url: string; group_claim?: string; group_mapping?: Record<string, string>; sso_required?: boolean }) =>
    apiClient.put<SsoConfig>(`${BASE}/sso`, data).then(r => r.data),
  deleteSsoConfig: () => apiClient.delete(`${BASE}/sso`),

  // Marketplace
  listMarketplace: () =>
    apiClient.get<MarketplacePlugin[]>(`${BASE}/marketplace`).then(r => r.data),
  seedMarketplace: () =>
    apiClient.post<MarketplacePlugin[]>(`${BASE}/marketplace/seed`).then(r => r.data),
  listInstalled: () =>
    apiClient.get<TenantPlugin[]>(`${BASE}/marketplace/installed`).then(r => r.data),
  installPlugin: (pluginId: string) =>
    apiClient.post<TenantPlugin>(`${BASE}/marketplace/${pluginId}/install`).then(r => r.data),
  uninstallPlugin: (pluginId: string) =>
    apiClient.delete(`${BASE}/marketplace/${pluginId}/uninstall`),

  // RBAC
  listRoles: () =>
    apiClient.get<RbacCustomRole[]>(`${BASE}/rbac-roles`).then(r => r.data),
  createRole: (data: { name: string; description?: string; permissions?: Record<string, unknown> }) =>
    apiClient.post<RbacCustomRole>(`${BASE}/rbac-roles`, data).then(r => r.data),
  deleteRole: (id: string) => apiClient.delete(`${BASE}/rbac-roles/${id}`),
  listAssignments: () =>
    apiClient.get<RbacRoleAssignment[]>(`${BASE}/rbac-assignments`).then(r => r.data),
  assignRole: (data: { user_id: string; role_id: string }) =>
    apiClient.post<RbacRoleAssignment>(`${BASE}/rbac-assignments`, data).then(r => r.data),
  removeAssignment: (id: string) => apiClient.delete(`${BASE}/rbac-assignments/${id}`),
};

File: backend/app/models/edge_agents.py
import uuid
from datetime import datetime
from typing import Optional, Any
from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class EdgeAgent(Base):
    __tablename__ = "edge_agents"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    token_hash: Mapped[str] = mapped_column(String(64), nullable=False, unique=True)
    device_ids: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)
    version: Mapped[Optional[str]] = mapped_column(String(30), nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="offline")
    last_seen: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    ip_address: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class SsoConfig(Base):
    __tablename__ = "sso_configs"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False, unique=True)
    provider: Mapped[str] = mapped_column(String(30), nullable=False, server_default="azure_ad")
    client_id: Mapped[str] = mapped_column(String(200), nullable=False)
    client_secret_encrypted: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    discovery_url: Mapped[str] = mapped_column(String(500), nullable=False)
    group_claim: Mapped[Optional[str]] = mapped_column(String(100), nullable=True, server_default="groups")
    group_mapping: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)
    extra_config: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)
    sso_required: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class MarketplacePlugin(Base):
    __tablename__ = "marketplace_plugins"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(200), nullable=False, unique=True)
    slug: Mapped[str] = mapped_column(String(100), nullable=False, unique=True)
    version: Mapped[str] = mapped_column(String(30), nullable=False)
    author_tenant_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    category: Mapped[str] = mapped_column(String(50), nullable=False, server_default="connector")
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    package_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    signature: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    approved_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    approved_by: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    is_builtin: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    download_count: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    tenant_installs: Mapped[list["TenantPlugin"]] = relationship("TenantPlugin", back_populates="plugin", cascade="all, delete-orphan")


class TenantPlugin(Base):
    __tablename__ = "tenant_plugins"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    plugin_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("marketplace_plugins.id", ondelete="CASCADE"), nullable=False)
    installed_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    installed_by: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    config: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)

    plugin: Mapped["MarketplacePlugin"] = relationship("MarketplacePlugin", back_populates="tenant_installs")


class RbacCustomRole(Base):
    __tablename__ = "rbac_custom_roles"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    permissions: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    assignments: Mapped[list["RbacRoleAssignment"]] = relationship("RbacRoleAssignment", back_populates="role", cascade="all, delete-orphan")


class RbacRoleAssignment(Base):
    __tablename__ = "rbac_role_assignments"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    role_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("rbac_custom_roles.id", ondelete="CASCADE"), nullable=False)
    assigned_by: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    assigned_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    role: Mapped["RbacCustomRole"] = relationship("RbacCustomRole", back_populates="assignments")


class SsoRoleMapping(Base):
    __tablename__ = "sso_role_mappings"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    sso_config_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("sso_configs.id", ondelete="CASCADE"), nullable=False, index=True)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False, index=True)
    external_group: Mapped[str] = mapped_column(String(300), nullable=False)
    platform_role: Mapped[str] = mapped_column(String(50), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
